File: editorial/chap6-21_25/prob6-25.cpp
#include <vector>
#include <iostream>
using namespace std;

int N, M, A[100009], B[100009]; vector<int> G[100009];

int dp[100009]; bool visited[100009];
void dfs(int pos) {
	visited[pos] = true;
	dp[pos] = 1;
	for (int i : G[pos]) {
		if (visited[i] == false) {
			dfs(i);
			dp[pos] += dp[i];
		}
	}
}

int main() {
	// 入力
	int N;
	cin >> N;
	for (int i = 1; i <= N - 1; ++i) {
		cin >> A[i] >> B[i];
		G[A[i]].push_back(B[i]);
		G[B[i]].push_back(A[i]);
	}
	
	// 深さ優先探索 (DFS) を使った動的計画法
	for (int i = 1; i <= N; i++) {
		visited[i] = false;
	}
	dfs(1);
	
	// 答えを計算して出力
	long long answer = 0;
	for (int i = 2; i <= N; i++) {
		answer += 1LL * dp[i] * (N - dp[i]);
	}
	cout << answer << endl;
	
	return 0;
}

File: editorial/chap6-21_25/prob6-25.py
import sys

# 深さ優先探索を行う関数
def dfs(pos, G, visited, dp):
	visited[pos] = True
	dp[pos] = 1
	for i in G[pos]:
		if visited[i] == False:
			dfs(i, G, visited, dp)
			dp[pos] += dp[i]

# 再帰呼び出しの深さの上限を 120000 に設定
sys.setrecursionlimit(120000)

# 入力
N = int(input())
A = [ None ] * (N - 1)
B = [ None ] * (N - 1)
for i in range(N - 1):
	A[i], B[i] = map(int, input().split())

# 隣接リストの作成
G = [ list() for i in range(N + 1) ]
for i in range(N - 1):
	G[A[i]].append(B[i])
	G[B[i]].append(A[i])

# 深さ優先探索 (DFS) を使った動的計画法
visited = [ False ] * (N + 1)
dp = [ None ] * (N + 1)
dfs(1, G, visited, dp)

# 答えを計算して出力
answer = 0
for i in range(2, N + 1):
	answer += dp[i] * (N - dp[i])
print(answer)